luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -49,29 +49,83 @@ local projectileDefIDCache = {}
 local shieldedUnits = {}
 local AOEWeaponDefIDs = {}
 local projectileShieldHitCache = {}
-local gameSeconds = 0
+local highestWeapDefDamages = {}
+local armoredUnitDefs = {}
 
 for weaponDefID, weaponDef in ipairs(WeaponDefs) do
-	if weaponDef.damageAreaOfEffect > aoeIgnoreThreshold and not weaponDef.customParams.shield_aoe_penetration then
+	if not weaponDef.customParams.shield_aoe_penetration then
 		AOEWeaponDefIDs[weaponDefID] = true
 	end
 
 	if weaponDef.customParams.beamtime_damage_reduction_multiplier then
 		local base = weaponDef.customParams.shield_damage or fallbackShieldDamage
 		local multiplier = weaponDef.customParams.beamtime_damage_reduction_multiplier
-		originalShieldDamages[weaponDefID] = math.ceil(base * multiplier)
+		originalShieldDamages[weaponDefID] = mathCeil(base * multiplier)
 	else
 		originalShieldDamages[weaponDefID] = weaponDef.customParams.shield_damage or fallbackShieldDamage
 	end
 
 	if weaponDef.type == 'Flame' or weaponDef.customParams.overpenetrate then
 		forceDeleteWeapons[weaponDefID] = weaponDef
 	end
+
+	local highestDamage = 0
+	if weaponDef.damages then
+		for type, damage in ipairs(weaponDef.damages) do
+			if damage > highestDamage then
+				highestDamage = damage
+			end
+		end
+	end
+
+	--this section calculates the rough amount of damage required to be considered a "direct hit", which assumes it didn't happen from AOE reaching inside shield.
+	local beamtimeReductionMultiplier = 1
+	local minIntensity = 1
+	if weaponDef.beamtime and weaponDef.beamtime < 1 then
+		local minimumMinIntensity = 0.5
+		local minIntensity = weaponDef.minIntensity or minimumMinIntensity
+		minIntensity = math.max(minIntensity, minimumMinIntensity)
+		-- This splits up the damage of hitscan weapons over the duration of beamtime, as each frame counts as a hit in ShieldPreDamaged() callin
+		-- Math.floor is used to sheer off the extra digits of the number of frames that the hits occur
+		beamtimeReductionMultiplier = 1 / math.floor(weaponDef.beamtime * Game.gameSpeed)
+   end
+
+   local minimumMinIntensity = 0.65
+   local damageFalloffUnitTypes = {
+	"BeamLaser",
+	"Flame",
+	"LaserCannon",
+	"LightningCannon"
+   }
+   local hasDamageFalloff = false
+   for _, exemption in ipairs(damageFalloffUnitTypes) do
+		if exemption == weaponDef.type then
+			hasDamageFalloff = true
+			break
+		end
+	end
```

Comment:
Don't need to iterate manually:
```suggestion
	local damageFalloffUnitTypes = {
		BeamLaser = true,
		Flame = true,
		LaserCannon = true,
		LightningCannon = true,
	}
	if damageFalloffUnitTypes[weaponDef.type] then
		hasDamageFalloff = true
	end
```

Reply:
yeah, that's way cleaner.